### PR TITLE
WIP:: LOG-1157 Add initial json log format test

### PR DESCRIPTION
**Pull request by eranra**
### Description
Initial test for json fomat
Implementation of the JSON forwarding proposal
https://github.com/openshift/enhancements/blob/master/enhancements/cluster-logging/forwarding-json-structured-logs.md

/cc @vimalk78 
/assign @jcantrill 

### Links
https://issues.redhat.com/browse/LOG-1157

### CODE REVIEW

File: test/functional/normalization/json_format_test.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+		Expect(framework.WritesJsonApplicationLogs(json, 1)).To(BeNil())
+	})
+	AfterEach(func() {
+		framework.Cleanup()
+	})
+
+	Context("when parsing json logs as un-structured", func() {
+		It("should be logged as a message", func() {
+			outputLogTemplate.Message = strings.Replace(json, "\"", "", -1)
+			raw, err := framework.ReadApplicationLogsFrom("fluentforward")
```
Comment (jcantrill) [inline test/functional/normalization/json_format_test.go:39]: I believe you can use the api type `logging.OutputFluentForward`

File: test/functional/normalization/json_format_test.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+		Expect(framework.Deploy()).To(BeNil())
+		Expect(framework.WritesJsonApplicationLogs(json, 1)).To(BeNil())
+	})
+	AfterEach(func() {
+		framework.Cleanup()
+	})
+
+	Context("when parsing json logs as un-structured", func() {
+		It("should be logged as a message", func() {
+			outputLogTemplate.Message = strings.Replace(json, "\"", "", -1)
```
Comment (jcantrill) [inline test/functional/normalization/json_format_test.go:38]: What exactly are we verifying here.  Initially as-is, the message field is a string representation of JSON.  What do we believe the message field looks like here?

**Review by jcantrill — COMMENTED** [review]
/approve

File: test/functional/normalization/json_format_test.go (hunk 2 above)
Comment (eranra) [reply test/functional/normalization/json_format_test.go]: @jcantrill you will be suprised .... what I saw is that we are getting an empty message ... and that the log message is added as Json key+value to the output ... not sure if anyone was planning on that ... anyway ... we will add more here over time ... let me change the text to call this WIP

File: test/functional/normalization/json_format_test.go (hunk 1 above)
Comment (eranra) [reply test/functional/normalization/json_format_test.go]: Done

File: test/functional/normalization/json_format_test.go (hunk 2 above)
Comment (eranra) [reply test/functional/normalization/json_format_test.go]: @jcantrill I updated the PR ... now if you look on option #2 you will see here: https://github.com/openshift/cluster-logging-operator/pull/932/files#diff-6bcd37f08b54c03e0e0a44f0cce637f8d3b13a3c15661aef973a76556d57e7f5R413 

The "leading" plan for the new behaiviour after API will be implmented for this feature ... this is the expected fluentd conf file (for now) This test compares current behviour and future behviour, after we will have support for json logs

Hope this make sense ???

Comment (eranra) [reply test/functional/normalization/json_format_test.go]: @alanconway ^^^ FYI

File: test/functional/normalization/fluentd_structured_configuration_option1.txt [hunk 3]
```diff
@@ -0,0 +1,453 @@
+## CLO GENERATED CONFIGURATION ###
```
Comment (jcantrill) [inline test/functional/normalization/fluentd_structured_configuration_option1.txt:1]: Please remove code from commit

File: test/functional/normalization/fluentd_structured_configuration_option2.txt [hunk 4]
```diff
@@ -0,0 +1,454 @@
+## CLO GENERATED CONFIGURATION ###
```
Comment (jcantrill) [inline test/functional/normalization/fluentd_structured_configuration_option2.txt:1]: please remove from commit

File: test/functional/normalization/fluentd_structured_configuration_option2.txt [hunk 5]
```diff
@@ -0,0 +1,454 @@
...
+      @type relabel
+      @label @FLUENTFORWARD
+    </store>
+  </match>
+</label>
+
+# Ship logs to specific outputs
+<label @FLUENTFORWARD>
+
+  ##### THIS IS NEW CONFIGURATION THAT WILL BE SET USING API
```
Comment (jcantrill) [inline test/functional/normalization/fluentd_structured_configuration_option2.txt:412]: We already have an embedded plugin which will parse JSON.  Additionally, this filter already exists to accomplish the same thing https://docs.fluentd.org/filter/parser

File: test/functional/normalization/json_format_test.go [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+			outputLogTemplate.Message = strings.Replace(jsonMsg, "\\", "", -1)
+			outputTestLog := logs[0]
+			Expect(outputTestLog).To(matchers.FitLogFormatTemplate(outputLogTemplate))
+		})
+	})
+
+	Context("when parsing json logs as structured - using current plugin implementation - option 1", func() {
+		It("should be logged as a message", func() {
+
+			framework.SetFluentConfigFileName("fluentd_structured_configuration_option1.txt")
```
Comment (jcantrill) [inline test/functional/normalization/json_format_test.go:57]: These configuration should come from logic to produce the config using the generator.  This test in particular should be added to the final implementation.

**Pull request closed without merging**